flowmachine/flowmachine/features/subscriber/mobility_classification.py

```diff
@@ -0,0 +1,124 @@
+      periods at any single location
+    - 'stable': Subscriber spent at least 'stay_length_threshold' consecutive
+      periods at the same location
+
+    Only subscribers appearing in the result of the reference location query
+    for the most recent period are included in the result of this query (i.e.
+    subscribers absent from the query result can be assumed to fall into a
+    sixth category: "not active in the most recent period").
+
+    Parameters
+    ----------
+    locations : list of BaseLocation
+        List of reference location queries, each returning a single location
+        per subscriber (or NULL location for subscribers that are active but
+        unlocatable). The list is assumed to be sorted into ascending
+        chronological order.
+    stay_length_threshold : int, default 3
+        Minimum number of consecutive periods over which a subscriber's
+        location must remain the same for that subscriber to be classified as
+        'stable'.
+    """
+
+    def __init__(self, locations: List[BaseLocation], stay_length_threshold: int = 3):
+        self.locations = locations
+        if len(set(l.spatial_unit for l in self.locations)) > 1:
+            raise InvalidSpatialUnitError(
+                "MobilityClassification requires all input locations to have the same spatial unit"
+            )
+        self.stay_length_threshold = int(stay_length_threshold)
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        return ["subscriber", "value"]
+
+    def _make_query(self) -> str:
+        loc_cols_string = ", ".join(self.locations[0].spatial_unit.location_id_columns)
+
+        # Workaround to count rows where all location columns are null, if there are multiple location columns
+        if len(self.locations[0].spatial_unit.location_id_columns) > 1:
+            count_unlocatable = f"sum((NOT (({loc_cols_string}) IS NULL))::int)"
+        else:
+            count_unlocatable = (
+                f"count({self.locations[0].spatial_unit.location_id_columns[0]})"
+            )
+
+        # Note: in some ways it would be nicer to use a DayTrajectories query instead of a list of location queries,
+        # but DayTrajectories only works with queries that have 'start' and 'stop' attributes
+        locations_union = " UNION ALL ".join(
+            f"SELECT subscriber, {loc_cols_string}, {i} AS ordinal FROM ({loc.get_query()}) _"
+            for i, loc in enumerate(self.locations)
+        )
+
+        long_term_activity = f"""
+        SELECT
+            subscriber,
+            count(*) < {len(self.locations)} AS sometimes_inactive,
+            {count_unlocatable} < {len(self.locations)} AS sometimes_unlocatable
+        FROM ({locations_union}) AS locations_union
+        GROUP BY subscriber
+        """
+
+        # Find stay lengths using gaps-and-islands approach
+        long_term_mobility = f"""
```

Comment:
I think I would consider breaking this in particular out into a separate query class.